lib/perl/Genome/Model/Tools/Bsmap/MethCalcConversionRate.pm

```diff
@@ -0,0 +1,263 @@
+            is => 'String',
+            is_optional => 1,
+            doc => 'Use genome model ID to calculate methylation conversion',
+        },
+        output_file => {
+            is => 'String',
+            is_optional => 1,
+            doc => 'Output methylation conversion',
+        },
+    ],
+};
+
+sub help_synopsis {
+  return <<EOS
+    gmt bsmap meth-calc-conversion-rate --model-id=394d6228a7b5487a9cb0ad0c448b5a44
+
+    gmt bsmap meth-calc-conversion-rate --snvs-file=/gscmnt/gc9016/info/model_data/394d6228a7b5487a9cb0ad0c448b5a44/buildf92b6057072948c2a6056a3ee412d596/variants/snv/meth-ratio-2.74-d41d8cd98f00b204e9800998ecf8427e/MT/snvs.hq
+
+EOS
+}
+
+sub help_brief {
+    "calculate methylation conversion using Mitochondrial DNA or spiked lambda"
+}
+
+sub help_detail {
+    "calculate methylation conversion using Mitochondrial DNA or spiked lambda"
+}
+
+sub bs_rate {
+    my $count = 0;
+    my $totalreads = 0;
+    my $methreads = 0;
+
+    open (my $fh, '<', $_[0]) or die("Could not open snvs file.");
+    while (my $line = <$fh>) {
+        if ( $count == 0 ) {
+            $count = 1;
+            next;
+        }
+        chomp($line);
+        my @F = split("\t",$line);
+        if(($F[2] eq "-" && $F[3] =~ /.CG../ ) || ($F[2] eq "+" && $F[3] =~ /..CG./)){
+            $totalreads = $totalreads + $F[5];
+            $methreads = $methreads + $F[6];
+        }
+    }
+    close($_[0]);
+
+    my $cfile;
+
+    if (defined($_[2])){
+        open($cfile, ">>$_[2]") || die "Can't open output file for writing.\n";
+        if($_[1] eq "MT"){
+            print $cfile "\nMethylation conversion based on mtDNA:\n";
+        }
+        if($_[1] eq "lambda"){
+            print $cfile "\nMethylation conversion based on lambda:\n";
+        }
+        print $cfile "Meth reads\t=\t", $methreads, "\n";
+        print $cfile "Total reads\t=\t", $totalreads, "\n";
+        if ($totalreads != 0) {
+            print $cfile "Bisulfite conversion (%)\t=\t", 100-($methreads/$totalreads*100), "\n\n";
+        }
+        close $cfile;
+    }
+    else {
+        print "Meth reads\t=\t", $methreads, "\n";
+        print "Total reads\t=\t", $totalreads, "\n";
+        if ($totalreads != 0) {
+            print "Bisulfite conversion (%)\t=\t", 100-($methreads/$totalreads*100), "\n\n";
+        }
+    }
+}
+
+sub execute {
+    my $self = shift;
+    my $snvs_file = $self->snvs_file;
+    my $model_id = $self->model_id;
+    my $output_file = $self->output_file;
+
+    if (!defined($snvs_file) && !defined($model_id) ){
+        die("Must provide snvs file OR model ID.\n");
+    }
+
+    if (defined($snvs_file) && defined($model_id) ){
+        die("Must provide snvs file OR model ID.\n");
+    }
+
+    my $cfile;
+
+    if (defined($snvs_file)){
+        # snvs
+        if (-s "$snvs_file"){
+            if (defined($output_file)){
+                open($cfile, ">$output_file") || die "Can't open output file for writing.\n";
```

Comment:
These two cases would also get de-duplicated if `bs_rate()` always wrote to a handle and default to a handle to stdout.